### ISSUE TITLE
Memory leak 🔍

### ISSUE THREAD

**Issue by alloy** [artsy/force#1199]
Follow-up to #1118.

Just going to dump my brain here as I look into things. Expect ramblings.

**Comment by alloy** [comment]
Looking at the diff between two heap dumps with the largest delta, reveals that the object taking up the most memory is `TraceSegment`. The next one is `Array`, but those are mostly being held by `TraceSegment` objects.

![screen shot 2017-04-10 at 15 21 41](https://cloud.githubusercontent.com/assets/2320/24863827/b3ac5d2a-1e02-11e7-9a29-edd8247ae1e7.png)

An example trace looks like the following:


* `transaction in TraceSegment @3271695` ([NewRelic](https://github.com/newrelic/node-newrelic/blob/v1.37.2/lib/transaction/trace/segment.js#L42))
* `__NR_parent in Layer @409349` ([NewRelic](https://github.com/newrelic/node-newrelic/blob/v1.37.2/lib/instrumentation/express.js#L361))
* `stack in router() @770155` ([Express](https://github.com/expressjs/express/blob/4.15.2/lib/router/index.js#L58))
* `_router in app() @770137` ([Express](https://github.com/expressjs/express/blob/4.15.2/lib/application.js#L139))
* `exports in Module @35677` ([Node.js](https://github.com/nodejs/node/blob/v6.10.2/lib/module.js#L39))
* `/app/desktop/apps/user/index.coffee in @5819` ([Force](https://github.com/artsy/force/blob/825664b92be4592dac5d97a68a402b449b4a2f79/desktop/apps/user/index.coffee#L4))
* `_cache in Module() @50457` ([Node.js](https://github.com/nodejs/node/blob/v6.10.2/lib/module.js#L51))

The Force frame in the trace isn’t limited to that specific file, I’ve seen various files (including `/apps/desktop/apps/style_guide/index.coffee` and `/apps/desktop/apps/dev/index.coffee`), but the trace pattern is always the same.

**Comment by alloy** [comment]
I haven’t spotted a way yet for those `TraceSegment` objects to be leaked. I’m wondering if these objects are simply for in-flight requests or else buffered to be sent to NewRelic shortly or might even be eligible for garbage collection but simply haven’t been collected yet 🤔

**Comment by alloy** [comment]
It seems like the `heapdump` module will trigger not 1, but even 2 garbage collections _before_ generating the actual snapshot:

1. https://github.com/bnoordhuis/node-heapdump/blob/7cf17e9b2a629fa5708717874ced2a7ae6cb2e13/src/heapdump.cc#L121
2. https://github.com/v8/v8/blob/5.1-lkgr/src/api.cc#L8521
3. https://github.com/v8/v8/blob/5.1-lkgr/src/profiler/heap-profiler.cc#L71
4. https://github.com/v8/v8/blob/5.1-lkgr/src/profiler/heap-snapshot-generator.cc#L2535-L2544

So seems like these are definitely not eligible for garbage collection.

**Comment by alloy** [comment]
I wonder if it makes sense to run Force for a while _without_ NewRelic enabled and see:

1. If the memory still grows out of control. (One thing I wonder is if the amount of requests we’re handling and the amount of memory tracing these requests takes isn’t slowing down the app to a grinding halt, inevitably starving.)
2. Or if that removes a lot of noise and allows us to drill down on the real leaks.

@craigspaeth What do you think?

**Comment by alloy** [comment]
Another interesting object is `ArtistResult`:

![screen shot 2017-04-10 at 16 15 25](https://cloud.githubusercontent.com/assets/2320/24865898/f96a7616-1e08-11e7-91a0-de65a6888a99.png)

* The `ArtistResult` objects are created [here](https://github.com/artsy/force/blob/170484506304fcf986f41bf08b07a1e389e07e12/mobile/models/fair.coffee#L105-L114)
* and are being populated by [this `fetchUntilEnd`](https://github.com/artsy/artsy-backbone-mixins/blob/6309f2a4d4e1a1458105d934b1ec89af6bb87e98/lib/fetch.coffee#L19-L37) function.
* That function generates a closure that [refers to itself](https://github.com/artsy/artsy-backbone-mixins/blob/6309f2a4d4e1a1458105d934b1ec89af6bb87e98/lib/fetch.coffee#L33).

I’m not familiar enough yet with Node/V8 to know if a closure referring to itself could introduce a circular reference and leak (?) or if it’s also just data that’s being retained by an in-flight request.

**Comment by craigspaeth** [comment]
👏  great investigation work @alloy ! I think you have to two solid leads here. I support toggling of NewRelic and measuring from there—a [`heroku config:unset NEW_RELIC_LICENSE_KEY `](https://github.com/artsy/artsy-newrelic/blob/master/index.js#L1) might be enough to do that today.

To the point about ArtistResult—I don't think I can confidently answer the circular reference point off the top of my head, but it is definitely likely that `fetchUntilEnd` is holding on to a significant amount of memory as it fills up the artists collection with each new page of data it crawls (however, one would expect that to be GCed after `fetchArtists` finishes). I noticed the stack in that screenshot drilling down into jade files and jade's template cache.... maybe Jade is actually holding on to `ArtistResult` objects for the A to Z page?

**Comment by izakp** [comment]
Nice work @alloy - that self-reference seems fine - it respects pagination of results, I suppose?  But the `remove: false` option seems suspicious to me.  We should definitely be looking for something in the new space that is not marked for GC, and this seems a likely candidate.

I would turn off NewRelic reporting after we have exhausted other options.

**Comment by izakp** [comment]
I'm seeing evidence of other instances of references to Jade templates retained in new space across other heap dump comparisons.  I think it would be worth investigating disabling or limiting Jade template caching.  This seems to corroborated by other modelResult objects, not just ArtistResult.

**Comment by alloy** [comment]
> that self-reference seems fine

@izakp Do you know what the rules are for V8’s GC around self-referencing closures?

>  We should definitely be looking for something in the new space that is not marked for GC, and this seems a likely candidate, given that # Deleted is 0.

It’s very likely that we’re just looking at data for an inflight request, in which case it indeed should not be freed yet.

> I would turn off NewRelic reporting after we have exhausted other options, it seems like some references are being properly GC'd (if i'm interpreting # deleted correctly)

Yeah, some are definitely being collected. To be clear, disabling NewRelic is **not** because I suspect that leaking, but removing its noise from the graph. So this isn’t about removing it to see if that solves the problem, I only just starting the process.

I’m disabling NewRelic for the rest of the day/night and will revisit tomorrow, in the hope that there will be new cleaner dumps.

**Comment by izakp** [comment]
@alloy re: removing noise - ok good call on that - go ahead and disable that

Re: the node VM's rules for GC around self-referencing closures... I am not aware of this in particular.  All I know about node's GC mechanism is that if objects are not moved from the "new space" to the "old space" when they are no longer referenced, they are not GC'd

**Comment by izakp** [comment]
@alloy just for context: I never set the node VM flag `--max_new_space_size` as I read that doing so might put the VM in a position where it could not allocate any new objects, and thus crash the node process.  Setting the parameter `--max_old_space_size` did not seem to have an effect on the memory growth, hence my hypothesis that it is not the old space that is the issue, but likely objects surviving in the new space too long.  I am not sure what causes objects / references to be "moved" from the new space to the old space... most articles I read says "they are moved after a while"

**Comment by alloy** [comment]
iirc they are moved after surviving 2 “scavenge” cycles of the GC, so not too long at all.

Where in the UI do you see if the objects live in new/old space? I’ve been meaning to identify long-lived objects? (Even though surviving 2 cycles is probably not long enough for what I wanted to know.)

**Comment by izakp** [comment]
@alloy okay - I think I was misunderstanding how the old space works - things are promoted after a certain number of sweeps of the new space (by scavenging) and long-lived references in the old space get GC'd by lazy sweeping.  I can't find anywhere in the UI an indication that the elements live in the new vs old space... I guess it's safe to assume we are dealing with old space for long-lived references, but the UI isn't helping me determine which they are

**Comment by izakp** [comment]
@alloy maybe it is as simple as the NR agent... https://nathanhoad.net/memory-leak-with-nodejs-on-heroku

Comparing 2 heapdumps from April 4th -ce476bbee34c46d1b82e1530d3403c43-20170405003013.heapsnapshot taken at 8:30pm and ce476bbee34c46d1b82e1530d3403c43-20170405030422.heapsnapshot taken at 11:05 pm and looking at the objects allocated between these two snapshots, NT Transaction, Trace and Timer are all appearing with a relatively low shallow size (actual size in of object memory) compared to their Retained size (references to other objects) - see https://developers.google.com/web/tools/chrome-devtools/memory-problems/memory-101

**Comment by alloy** [comment]
So memory consumption has been largely stable, the slight ramp that can currently be seen may just be all the lazy CoffeeScript/Jade compilation.

Thus far I’m not convinced, however, that NewRelic might be leaking, too many of those articles (like the one you linked to) provide no proof at all that there was an actual leak. I want to be able to reproduce the issue, properly understand what’s going on, and implement a fix based on that.

I suggest we led Force run another day or two without NewRelic and see if it starts climbing again. If that happens, I’m also interested in seeing if the timeouts we’re seeing might lead to resources not being cleaned-up or something.

**Comment by alloy** [comment]
Quoting [myself from Slack](https://artsy.slack.com/archives/C0350895E/p1491834411847665) for posterity sake:

> Yeah I looked into a few discussed on their support forums, but we are either not on those versions or the problems being discussed appeared to be reported by people that didn’t really know how to debug the issues you were having, so big chance those are red herrings.
>
> I’m slightly afraid that this might be another case of unbound concurrency, like with MP, where the amount of concurrent requests we accept + the amount of memory each request uses + the app being too busy to run GC + limits imposed by Heroku lead to the app eventually being killed.
>
> It’s interesting how node appears to be such a double edged sword. Too much concurrency, what could go wrong?
>
> Anyways, would be interesting to see what the memory graph looks like _without_ NewRelic adding noise.

**Comment by alloy** [comment]
So memory still seems stable on day 2 after removing NewRelic ¯\_(ツ)_/¯

For other reasons, as detailed in #1033, we’re going to be trying other (error reporting) services anyways, so for now I won’t go and try to profile/debug NewRelic’s node agent to try and find potential leaks.

However, I want to be clear (especially for others potentially reading along), this should not be taken as a sign that NewRelic’s node agent was indeed leaking, the exact cause is still undetermined and for now I’m leaning towards the high number of concurrent inflight requests, combined with large amount of data NR collects being the issue, and Heroku being rather limited in memory allowance.

**Comment by izakp** [comment]
@alloy I wonder if upgrading the newrelic agent might solve the issue https://docs.newrelic.com/docs/agents/nodejs-agent/troubleshooting/troubleshooting-large-memory-usage-nodejs says to go to 1.18.0 or higher, where we have it packaged at 1.0.1 here: https://github.com/artsy/artsy-newrelic/blob/master/yarn.lock#L40

**Comment by izakp** [comment]
I would rather pursue this solution than do all the work to move everything to another instrumentation / error reporting service.

**Comment by cavvia** [comment]
From a platform perspective we'd like to get some basic performance monitoring back up as a priority. Advanced error reporting sounds fine and can happen in parallel maybe, but it's a different kettle of fish altogether.

A stop-gap measure might be to add some Heroku alerting on response times and throughput, which are available to us as we use professional dynos - I can do that now in fact - but it's not a long term solution.

**Comment by alloy** [comment]
@izakp Force is on version 1.37.2: https://github.com/artsy/force/blob/ac24a3c2ef5db90c23cdee815ee20d90a9dd3690/yarn.lock#L4929, which isn’t the latest but not old either and there don’t appear to have been any related fixes since then https://github.com/newrelic/node-newrelic/blob/master/NEWS.md. We should still upgrade, but I don’t expect it to improve this situation.

**Comment by alloy** [comment]
> I would rather pursue this solution than do all the work to move everything to another instrumentation / error reporting service.

We can spend a lot of time on hunting down the exact issue, but we should really consider if in the longer-term we would actually keep using NewRelic before investing that time.

> A stop-gap measure might be to add some Heroku alerting on response times and throughput, which are available to us as we use professional dynos - I can do that now in fact - but it's not a long term solution.

I would prefer that, if you want to enable something asap.

**Comment by cavvia** [comment]
I've added alerts on response times and high error counts, which is all heroku will allow. It gives us basic performance monitoring though.

**Comment by kanaabe** [comment]
Fwiw I did a lot of firefighting back in March on Positron when we changed a version of New Relic and experienced memory leaks/app crashes. The cause was an incompatible version of New Relic with the version of node we have. Doesn't seem exactly related but maybe helpful.

Described here: https://github.com/artsy/positron/pull/966

**Comment by izakp** [comment]
@craigspaeth @alloy @sweir27 any plans to, as @craigspaeth mentioned, Upgrade to Node 7 and reinstate NewRelic?

**Comment by alloy** [comment]
We can, but, as I’ve been trying to say, I don’t think there was a memory leak at all, just New Relic using a lot of memory and that combined with another reason that leads to many in-flight requests being an issue.

**Comment by izakp** [comment]
@alloy i'm going to take another look at the hap snapshots but I am quite certain the transactions I looked at were unresolved weak refs... without evidence to the contrary, and what with the reported bugs in the newrelic agent... I still think this was evidence of a memory leak

**Comment by izakp** [comment]
Ok so just to recap (Again I should have grabbed screenshots while I saw this behavior) but what I observed was a gradual increase (over the span of 12-24 hours) of memory consumption, starting at   ~500Mb (where the app is now, and generally, under normal load) and then growing linearly, unbounded, until the 2Gb limit, where Heroku killed the dyno.  That behavior does not, in my judgement, correlate to in-flight requests, as these manifest in transient spikes.  So i'm really at an impasse as for how to continue further without even further instrumentation.
